### DISCUSSION TITLE
Opendds_idl and Tao_idl is not generated for Target build

### DISCUSSION THREAD

**Discussion by pmaskale** [OpenDDS/OpenDDS#3562] [Q&A] [answered]
I have built opendds with cross compiler for aarch64. 
Is there a reason why opendds_idl and tao_idl is not generated for target? Though it generates for host.

Is there a way to generate these for Target (aarch64)?

when i run my cmake in cross compiler environment : OPENDDS_TARGET_SOURCES is not found.
The reason being opendds_idl is x86 compiled and is not understood by aarch64.

workaround : I will rerun the cmake required for generating in x86 which understands(OPENDDS_TARGET_SOURCES) and later build with aaarch64.

Is there a clean solution for this?

**Comment by mitza-oci** [comment, marked as answer]
The code generators are not used on the target.  Compiling happens in the host environment.

Note that the use of the term "target" in CMake's `TARGET_SOURCES` is not related to the cross-compiler target.

**Reply by iguessthislldo** [reply]
It's also important to note that I don't think we support cross compiling with CMake yet. I've done some experiments with Android Studio, but I didn't get far enough to get it working. I'm not sure it needs much to get it going in a simple cross compile scenario though, probably just a few extra checks to change how it finds the IDL compilers.

**Reply by pmaskale** [reply]
Thank you for the responses. 
Nevertheless i was able to run my application with opendds by building on host with cross compiler using cmake with the workaround.
@iguessthislldo do you mean finding of x86 architecture generators (opendds_idl) from aarch64 cross compiler environment through cmake? 
@mitza-oci  Then always it is like generate the file in supporting platform(x86) and take the generated code and push to the cross compile environment to build?

**Reply by mitza-oci** [reply]
It is probably easiest to work from an example.  You've already built OpenDDS in a cross-compile configuration.  It has two subdirectories of interest: `build/host` which creates programs that run "on the host" (for example a Linux laptop / x86_64) and `build/target` which creates programs that run "on the target" (for example a Raspberry Pi / ARM).  Within `build/target` you'll find `DevGuideExamples/DCPS/Messenger` which contains an IDL file.  When this project is built, the code generators that were built in `build/host` are run to generate code from IDL.  Then that code is compiled with the cross-compiler along with the rest of the code for this project.  Note that this doesn't use CMake.
